### DISCUSSION TITLE
Suppress warnings in promises

### DISCUSSION THREAD

**Discussion by larsewi** [cfengine/core#4508] [Q&A] [answered]
Is there a way to suppress warnings in promises? If not, should it be a feature?
This would be helpful to reduce noise from warnings you don't care about.

e.g.
```
bundle agent my_bundle
{
    somepromise: 
        "foo" 
            bar => "baz",
            supress_warning => "foobar"; # we know about this warning, but we simply don't care
}
```

**Comment by olehermanse** [comment]
@larsewi Interesting, do you have any examples of warnings you'd like to suppress?

**Reply by larsewi** [reply]
Honestly, I can't think of any examples. This is probably due to the fact that CFEngine – to my experience – rarely logs warnings. Why? I would guess to reduce noise.

E.g.
```
users:
  "123"
    policy => "present";
```
username `123` is not according to the systems reference manuals recommendation _(see useradd(8))_, thus I would expect a warning, but I don't get one. 

But then again, if we cannot suppress it, we're better of without it. Imagine some company with hundreds of users, and they have some weird username scheme. That would be quite noisy.

**Comment by nickanderson** [comment, marked as answer]
No, I don't believe such functionality currently exists, but I can imagine the usefulness.

The closest thing that I know of are the mailfilter_include and mailfilter_exclude executor control attributes. You can define them as regular expressions and they will use that to filter lines for the emails cf-execd sends.

My first thought is that suppressing some specific style warnings would be done in body common control.

Perhaps it would be most useful as something very generic. For example, log_include, log_exclude that applies to the output of all daemons, not just emails sent by cf-execd as described above.

**Reply by larsewi** [reply]
In Java you have
```Java
@SuppressWarnings({"unchecked", "deprecated"})
```
where you suppress types of warnings within this scope, which is quite useful sometimes. 

Could for instance `suppress_warnings` be a promise type, that resides within a bundle, just like the `vars` promise type?

**Reply by olehermanse** [reply]
I would expect to be able to control this from command line arguments and inline comments or macros. That's how most compilers and programming languages handle it, AFAIK.

@larsewi I guess you should accept @nickanderson's answer - no it's not currently possible.

I don't think it's really necessary at this point, I don't know of any situations where it's needed. If we do a more advanced linter / syntax checker, it will definitely be necessary, and it will probably be implemented using command line arguments and comments. Just like cppcheck, LGTM, clang-whatever etc.

**Reply by larsewi** [reply]
Thank you @nickanderson & @olehermanse for your answers!
